saneterm/color.py

```diff
@@ -0,0 +1,158 @@
+      integers in the range [0;255].
+    """
+    def __init__(self, t, data):
+        if not isinstance(t, ColorType):
+            raise TypeError("type must be ColorType")
+
+        if t is ColorType.TRUECOLOR:
+            if not type(data) is tuple:
+                raise TypeError("data must be tuple for TRUECOLOR")
+            if not len(data) == 3:
+                raise TypeError("tuple must have 3 elements for TRUECOLOR")
+        elif t is ColorType.NUMBERED_8 or t is ColorType.NUMBERED_8_BRIGHT:
+            if not isinstance(data, BasicColor):
+                raise TypeError(f'data must be BasicColor for {t}')
+        elif t is ColorType.NUMBERED_256:
+            if not type(data) is int:
+                raise TypeError('data must be integer for NUMBERED_256')
+            if not (data >= 0 and data < 256):
+                raise TypeError('data must be in range [0;255] for NUMBERED_256')
+
+        self.type = t
+        self.data = data
+
+    def to_gdk(self):
+        """
+        Convert a Color into a Gdk.RGBA which TextTag accepts.
+        The color scheme for the 16 color part uses default X11
+        colors and is currently not configurable.
+        """
+        if self.type is ColorType.NUMBERED_8:
+            return basic_color_to_rgba(self.data, bright=False)
+        elif self.type is ColorType.NUMBERED_8_BRIGHT:
+            return basic_color_to_rgba(self.data, bright=True)
+        elif self.type is ColorType.TRUECOLOR:
+            return int_triple_to_rgba(self.data)
+        elif self.type is ColorType.NUMBERED_256:
+            if self.data < 8:
+                # normal 8 colors
+                return basic_color_to_rgba(BasicColor(self.data), bright=False)
+            elif self.data >= 8 and self.data < 16:
+                # bright 8 colors
+                return basic_color_to_rgba(BasicColor(self.data - 8), bright=True)
+            elif self.data >= 16 and self.data < 232:
+                # color cube which is constructed in the following manner:
+                #
+                # * The color number is described by the following formula:
+                #   n = 16 + 36r + 6g + b
+                # * r, g, b are all >= 0 and < 6
+                # * The corresponding color channel value for the r, g, b
+                #   values can be obtained using the following expression:
+                #   x * 40 + 55 if x > 0 else 0
+                #
+                # This is not documented anywhere as far as I am aware.
+                # The information presented here has been reverse engineered
+                # from XTerm's 256colres.pl.
+                tmp = self.data - 16
+                (r, tmp) = divmod(tmp, 36)
+                (g, b) = divmod(tmp, 6)
+
+                triple = tuple(map(extended_color_val, (r, g, b)))
+                return Gdk.RGBA(*triple)
+            else:
+                # grayscale in 24 steps
+                c = (self.data - 232) * (1.0/24)
```

Comment:
Maybe use constants for these magic numbers to improve readability? Briefly skimming over this it is not entirely clear to me why `self.data` is substracted by `232`.

Reply:
Is it better now?